### PR TITLE
Implement gitpod auth and settings sync

### PR DESCRIPTION
**Pull request by jeanp413**
related https://github.com/gitpod-io/gitpod/issues/9136

##### Common properties:
```
'common.extname': 'gitpod.gitpod-desktop',
'common.extversion': '0.0.29',
'common.nodeArch': 'x64',
'common.os': 'linux',
'common.platformversion': '5.13.0',
'common.product': 'desktop',
'common.uikind': 'desktop',
'common.vscodemachineid': '0d34d9a02c8981a664b756b27f73587b165f6c17b020744a3f2f5167081e1682',
'common.vscodesessionid': '859abcd9-0bb7-4b3d-a33c-574ac0ff2a631650380562678',
'common.vscodeversion': '1.67.0-insider',
```

- **gitpod_desktop_auth**
  - Properties:
    - **kind**: `login` or `logout` or `login_cancelled` or `login_failed` or `logout_failed`
    - **scopes**: `'["function:accessCodeSyncStorage","function:getGitpodTokenScopes","function:getLoggedInUser","resource:default"]'`
- **gitpod_desktop_settings_sync**
  - Properties:
    - **enabled**: `'true'` or `'false'`
- **gitpod_desktop_installation**
  - Properties:
    - **kind**: `install`

[gitpod-desktop-0.0.37.zip](https://github.com/gitpod-io/openvscode-server/files/8575373/gitpod-desktop-0.0.37.zip)

### CODE REVIEW

**Comment by akosyakov** [comment]
I think we can do analytics with another PR?

**Comment by akosyakov** [comment]
Screenshot 2022-04-14 at 09 50 35

Let's change it to:
- `Restart VS Code for the new Setting Sync configuration take effect.`
- `Reload this window`

File: extensions/gitpod/src/gitpodServer.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+import * as vscode from 'vscode';
+import { Log } from './common/logger';
+import fetch from 'node-fetch';
+import { PromiseAdapter, promiseFromEvent } from './common/utils';
+import { withServerApi } from './internalApi';
+import pkceChallenge from 'pkce-challenge';
+import { v4 as uuid } from 'uuid';
+import { Disposable } from './common/dispose';
+
+export const baseURL = 'https://jp-oauth-test.staging.gitpod-dev.com';
```
Comment (akosyakov) [inline extensions/gitpod/src/gitpodServer.ts:14]: it should be configurable via `gitpod.host`, i.e. `https://${gitpod.host}` on user level

File: extensions/gitpod/src/gitpodServer.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+export interface IGitpodServer extends vscode.Disposable {
```
Comment (akosyakov) [inline extensions/gitpod/src/gitpodServer.ts:16]: (nit) do we need this indirection?

**Comment by akosyakov** [comment]
Wow, code looks so well-structured ❤️ and you seem to address all important concerns.

File: extensions/gitpod/package.json [hunk 3]
```diff
@@ -29,8 +29,39 @@
...
+					"type": "object"
+				}
+			}
+		},
+		"commands": [
+			{
+				"command": "gitpod.auth.remove",
+				"category": "Gitpod",
+				"enablement": "gitpod.addedSyncProvider",
+				"title": "Disable Gitpod Settings Sync Provider"
```
Comment (akosyakov) [inline extensions/gitpod/package.json:55]: Is it surfaced somewhere on UI? It sounds very technical, but I could not find it on UI. can we remove `title` property to avoid confusion?

File: extensions/gitpod/src/gitpodServer.ts (hunk 2 above)
Comment (jeanp413) [reply extensions/gitpod/src/gitpodServer.ts:16]: removed it

File: extensions/gitpod/package.json (hunk 3 above)
Comment (jeanp413) [reply extensions/gitpod/package.json:55]: it's in the command pallete

**Comment by jeanp413** [comment]
> 
> Let's change it to:
> 
> * `Restart VS Code for the new Setting Sync configuration take effect.`
> * `Reload this window`

reload it not enough, you need to close all windows

**Comment by jeanp413** [comment]
> I think we can do analytics with another PR?

telemetry is done just needed a few :lipstick: I'll upload another vsix

File: extensions/gitpod/src/gitpodServer.ts (hunk 1 above)
Comment (jeanp413) [reply extensions/gitpod/src/gitpodServer.ts:14]: what do you mean another another `configuration` in package.json?

**Comment by jeanp413** [comment]
Added vsix file updated with telemetry (telemetry key added manually before creating the vsix for now, asked @jakobhero if it's Ok to push it to github or not, MS does commit it in it's extensions)

File: extensions/gitpod/src/gitpodServer.ts (hunk 1 above)
Comment (akosyakov) [reply extensions/gitpod/src/gitpodServer.ts:14]: yes

**Comment by akosyakov** [comment]
> reload it not enough, you need to close all windows

yes, that's why I wrote `Restart` in message and `reload this window`. As far as I understand `close this window` does not close it but reload?

File: extensions/gitpod/package.json (hunk 3 above)
Comment (akosyakov) [reply extensions/gitpod/package.json:55]: I thought it is `Gitpod: Turn Setting Sync On/Off`?

Comment (jeanp413) [reply extensions/gitpod/package.json:55]: I changed it because it's misleading, it doesn't turn settings sync at all, for that user has to do it manually by invoking `Settings Sync: Turn On...` command or in the menu option

File: extensions/gitpod/src/gitpodServer.ts (hunk 1 above)
Comment (jeanp413) [reply extensions/gitpod/src/gitpodServer.ts:14]: Added `gitpod.serviceUrl` configuration

**Comment by jeanp413** [comment]
@akosyakov added event description at the PR description

**Comment by akosyakov** [comment]
It seems that code validating token was dropped, why? We should make sure that if a token gets expired on server side or we modify an extension to add more scopes then token should be invalidated.

**Comment by akosyakov** [comment]
@jeanp413 @filiptronicek Could you rephrase UI facing messages to avoid `Gitpod Code Sync Provider`. It should be called `Setting Sync with Gitpod`. I see user facing error messages like `Error setting up Gitpod Code Sync Provider` and so on. Please scan source code of destkop extension and replace it with talking about enabling/disabling Settings Sync with Gitpod.

File: extensions/gitpod/src/extension.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		if(e.affectsConfiguration('gitpod')){
+			updateSyncContext();
+		}
+	}));
+
+	context.subscriptions.push(vscode.commands.registerCommand('gitpod.syncProvider.remove', async () => {
+		try {
+			await enableSettingsSync(false, telemetry);
+		} catch (e) {
+			vscode.window.showErrorMessage(`Error setting up Gitpod Code Sync Provider: ${e}`);
```
Comment (akosyakov) [inline extensions/gitpod/src/extension.ts:36]: Please log errors to output that they all captured to logs. Please review other try-catch in the desktop extension.

File: extensions/gitpod/package.json (hunk 3 above)
Comment (akosyakov) [reply extensions/gitpod/package.json:70]: I don't think anyone will understand what `Gitpod Settings Sync Provider` means either.

Can we contribute following items instead:
- `Settings Sync: Enable signing in with Gitpod account`
- `Settings Sync: Disable signing in with Gitpod account`

And fix error messaging correspondingly to something `Failed to enable singing in with Gitpod account: ${e}` or `Failed to disable singing in with Gitpod account: ${e}`

File: extensions/gitpod/src/authentication.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+			return [];
+		}
+
+		const sessionPromises = sessionData.map(async (session: SessionData) => {
+			// For the Gitpod scope list, order doesn't matter so we immediately sort the scopes
+			const sortedScopes = session.scopes.sort();
+			const scopesStr = sortedScopes.join(' ');
+
+			let userInfo: { id: string; accountName: string } | undefined;
+			if (!session.account) {
```
Comment (akosyakov) [inline extensions/gitpod/src/authentication.ts:125]: We need to validate here that token still has scopes which are required by current version of extension. It will bit as next time when we add a new scope, but functionallity won't work since old token will be used without invalidation. It should be common pattern in all our clients.

File: extensions/gitpod/src/authentication.ts [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+	private _sessionsPromise: Promise<vscode.AuthenticationSession[]>;
+
+	constructor(private readonly context: vscode.ExtensionContext, logger: Log, telemetry: TelemetryReporter) {
+		super();
+
+		this._logger = logger;
+		this._telemetry = telemetry;
+		this._gitpodServer = new GitpodServer(this._logger);
+		this._keychain = new Keychain(this.context, `gitpod.auth`, this._logger);
```
Comment (akosyakov) [inline extensions/gitpod/src/authentication.ts:40]: it should be scoped per gitpod host, if `gitpod.host` setting is changed then we should reread sessions for given host and unload for old host.

File: extensions/gitpod/package.json [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+			{
+				"label": "Gitpod",
+				"id": "gitpod"
+			}
+		],
+		"configuration": [
+			{
+				"title": "Gitpod",
+				"properties": {
+					"gitpod.serviceUrl": {
```
Comment (akosyakov) [inline extensions/gitpod/package.json:46]: why not just `host`? It is not only about URIs, but we scope all data of extension to some host and if this setting is chagned then we should unload data for old host and load for new host

**Comment by akosyakov** [comment]
@jeanp413 It looks like we will produce a lot of tables with such analytics events. Can we rather have only 3 events with different properties and timestamp to indicate progress:
- `gitpod_desktop_installation` to track when extension is installed/uninstalled
- `gitpod_desktop_settings_sync` to track enablement 
- `gitpod_desktop_auth` to track progress with Gitpod auth

These events thought won't be enough still to understand whether a user is really synching with us. Is there a way to track it from VS Code extension or we need to track in server then? machine ids are propagated with sync events as headers?

File: extensions/gitpod/package.json (hunk 7 above)
Comment (filiptronicek) [reply extensions/gitpod/package.json:46]: I agree `host` makes more sense here

File: extensions/gitpod/src/authentication.ts (hunk 6 above)
Comment (filiptronicek) [reply extensions/gitpod/src/authentication.ts]: I have now changed this so the key for the keychain is `gitpod.auth.host`.

File: extensions/gitpod/package.json (hunk 7 above)
Comment (filiptronicek) [reply extensions/gitpod/package.json]: Changed it in https://github.com/gitpod-io/openvscode-server/pull/337/commits/606c2c5198a5321ad612581ef056b9dcd06e45f7, I think it could be written in a clearer way, but I think just inputting the hostname is better than a URL, since it is shorter and makes sense to have in some places (logging, keychain keys).

I am open to reverting it back to the URL inputting though, we could do some easy parsing.

File: extensions/gitpod/src/extension.ts (hunk 4 above)
Comment (filiptronicek) [reply extensions/gitpod/src/extension.ts]: Checked them all out and it seems double-logging (user facing and to output) are everywhere. Made some minor tweaks in e746a4841d10d2a255727fcf4333def8fe2faa47.

File: extensions/gitpod/package.json (hunk 3 above)
Comment (filiptronicek) [reply extensions/gitpod/package.json:70]: I like @akosyakov's idea, but I am not sure if we can have the `Settings Sync:` prefix, since the extension is named `Gitpod` and so all commands start with that. This means that the command would be `Gitpod: Settings Sync: Enable signing in with Gitpod account`.
To be honest I am not sure what to change this to, but perhaps something like `Gitpod: Sign In to VS Code`? It is quite difficult to make this text short and concise, it is important that we tell the user that:
a) this is only useful for settings sync
b) this does not explicitly enable it, but it is the first step

File: extensions/gitpod/package.json (hunk 7 above)
Comment (jeanp413) [reply extensions/gitpod/package.json]: I'm fine with the rename (forgot it was suggested in the first place) but I'd much prefer to be a URL as before, that way I can just copy it form the browser and not think about it, now it will throw an error

Comment (filiptronicek) [reply extensions/gitpod/package.json]: Thanks for the feedback, @jeanp413. I'll revert it back to a full URL.

Comment (filiptronicek) [reply extensions/gitpod/package.json]: https://github.com/gitpod-io/openvscode-server/pull/337/commits/71b6ac84091c368dbd00adbe1f4311f0b8cb9ddf

File: extensions/gitpod/src/authentication.ts (hunk 5 above)
Comment (jeanp413) [reply extensions/gitpod/src/authentication.ts:126]: We don't need to validate scopes, scopes and token are 1 to 1 and stored in db, if we add/remove a scope in a new version of the extension then `getSessions` will be called with the new required scopes and it will filter from all the sessions returning only the ones which meets the required scopes.

**Comment by jeanp413** [comment]
> `gitpod_desktop_installation` to track when extension is installed/uninstalled

I think there's no API to know from within the extension when it was installed the first time or uninstalled

> These events thought won't be enough still to understand whether a user is really synching with us. Is there a way to track it from VS Code extension or we need to track in server then? machine ids are propagated with sync requests as headers?

We'll need to do it in the server, no API for that

File: extensions/gitpod/package.json (hunk 3 above)
Comment (akosyakov) [reply extensions/gitpod/package.json:70]: You can drop category and just provide title: `Settings Sync: Enable signing in with Gitpod account`.

Comment (filiptronicek) [reply extensions/gitpod/package.json:70]: I think we can make it shorter and drop the **account** there (I believe it's not necessary), but otherwise that sounds good 👍

Comment (jeanp413) [reply extensions/gitpod/package.json]: done

**Comment by jeanp413** [comment]
Added `gitpod_desktop_installation` event on first install

**Review by filiptronicek — APPROVED** [review]
LGTM 🎉 
image

File: extensions/gitpod/src/authentication.ts [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+		this._gitpodServer = new GitpodServer(serviceUrl, this._logger);
+		this._register(vscode.workspace.onDidChangeConfiguration(async e => {
+			if (e.affectsConfiguration('gitpod.host')) {
+				const serviceUrl = vscode.workspace.getConfiguration('gitpod').get<string>('host')!;
+				this._gitpodServer.dispose();
+				this._gitpodServer = new GitpodServer(serviceUrl, this._logger);
+
+				// Remove all sessions when gitpod.host changes
+				const sessions = await this._sessionsPromise;
+				await this._keychain.deleteToken();
```
Comment (akosyakov) [inline extensions/gitpod/src/authentication.ts:51]: I'm not sure that it works properly.

When a user updates `host` we should delete session from keychain for old host? But only unload it from memory and load other to avoid auth again when a user switch back. So I will imagine that `Keychain` object accept `host` for each operation.

Could you confirm that switching between hosts does not require unnecessary auth again with current implementation?

**Review by akosyakov — APPROVED** [review]
it looks good to me, besides https://github.com/gitpod-io/openvscode-server/pull/337#discussion_r859574191

@filiptronicek Could you double check following:
- switching between Gitpod host does not require auth again because we delete token unnecessary
- opening with VS Code Desktop still works

File: extensions/gitpod/src/authentication.ts (hunk 8 above)
Comment (jeanp413) [reply extensions/gitpod/src/authentication.ts:51]: It requires auth again, why would users keep changing host? either they use saas or self hosted, changing host is such a cumbersome operation because you need to restart vscode (close all windows) again for it to take effect
And then it would be misleading you could enable setting sync with one host then change the host and settings sync will keep working and you would think it using the new host

Comment (akosyakov) [reply extensions/gitpod/src/authentication.ts:51]: Yes, switch between gitpod.io and self hosted installation. It is a rare operation and then indeed we need to restart to apply changes. But I don't think annoy with requesting token each time is a way.

I wonder why is it so hard to implement properly? We did the same for local app and JetBrains and there a user can switch nicely for testing and self hosted. Let'd do proper here too?

Comment (jeanp413) [reply extensions/gitpod/src/authentication.ts:51]: will take another

**Comment by jeanp413** [comment]
@filiptronicek can you test and if everything looks good to you too approve it so we can merge this and release :rocket: 
- switching between Gitpod host does not require auth again because we delete token unnecessary

File: extensions/gitpod/src/authentication.ts [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+		const sessions = await this._sessionsPromise;
+		const finalSessions = sortedScopes.length
+			? sessions.filter(session => arrayEquals([...session.scopes].sort(), sortedScopes))
+			: sessions;
+
+		this._logger.info(`Got ${finalSessions.length} sessions for ${sortedScopes?.join(',') ?? 'all scopes'}...`);
+		return finalSessions;
+	}
+
+	private async checkForUpdates() {
```
Comment (akosyakov) [inline extensions/gitpod/src/authentication.ts:80]: probably it is minor but this code can race, it would be better to make it cancellabe

File: extensions/gitpod/package.json [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+					}
+				}
+			}
+		],
+		"commands": [
+			{
+				"command": "gitpod.syncProvider.remove",
+				"category": "Settings Sync",
+				"enablement": "gitpod.addedSyncProvider",
+				"title": "Disable signing in with Gitpod"
```
Comment (akosyakov) [inline extensions/gitpod/package.json:70]: `Disable Sign In with Gitpod`

File: extensions/gitpod/package.json [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+			},
+			{
+				"command": "gitpod.syncProvider.add",
+				"category": "Settings Sync",
+				"enablement": "!gitpod.addedSyncProvider",
+				"title": "Enable signing in with Gitpod"
```
Comment (akosyakov) [inline extensions/gitpod/package.json:76]: `Enable Sign In with Gitpod`

**Review by akosyakov — APPROVED** [review]
It worked nicely for me. I left some comments about user facing naming casing. It does not match to how other commands are named.

**Review by filiptronicek — APPROVED** [review]
> opening with VS Code Desktop still works

Yes it does! All looking good.

For some reason I also wanted to try installing extensions not on OpenVSX. It looks like it is behaving correctly by not removing it, just skipping it from sync in places where it is not available...

```
[2022-04-28 18:47:33.081] [settingssync] [info] Extensions: Skipped synchronizing extension because the extension is not found. github.copilot
```

File: extensions/gitpod/src/authentication.ts (hunk 9 above)
Comment (jeanp413) [reply extensions/gitpod/src/authentication.ts:80]: in theory yes, but I think it's really hard or impossible to happen as it depends on user input and the `this._sessionsPromise` is most probably already resolved, making it cancellable will make it race too in the second await, I think a queue will "fix" it, but I'll leave it for later
